torchao/prototype/quantized_training/int8_mixed_precision.py

```diff
@@ -0,0 +1,231 @@
+
+    # workaround for FSDP2 with `MixedPrecisionPolicy(param_dtype)`
+    # see `Int8MixedPrecisionTrainingLinearWeight.fsdp_pre_all_gather()` for more details.
+    fsdp_param_dtype: Optional[torch.dtype] = None
+
+
+_DEFAULT_CONFIG = Int8MixedPrecisionTrainingConfig()
+
+
+aten = torch.ops.aten
+
+
+class Int8MixedPrecisionTrainingLinearWeight(TorchAOBaseTensor):
+    """Linear weight for INT8 mixed-precision training. The weight is in original precision (e.g. FP32 or BF16).
+    During training, weight and activation are dynamically quantized and cast to INT8 to utilize INT8 Tensor Cores,
+    and then scaled back to original precision. This is also applied to backward pass.
+    """
+
+    @staticmethod
+    @torch._dynamo.disable
+    def __new__(cls, data: Tensor, config: Int8MixedPrecisionTrainingConfig):
+        return Tensor._make_wrapper_subclass(
+            cls,
+            data.shape,
+            data.stride(),
+            data.storage_offset(),
+            dtype=data.dtype,
+            device=data.device,
+        )
+
+    @torch._dynamo.disable
+    def __init__(self, data: Tensor, config: Int8MixedPrecisionTrainingConfig):
+        self._data = data
+        self.config = config
+
+    def __tensor_flatten__(self):
+        return ["_data"], [self.config]
+
+    @classmethod
+    def __tensor_unflatten__(cls, tensor_data_dict, tensor_attributes, outer_size=None, outer_stride=None):
+        return cls(tensor_data_dict["_data"], *tensor_attributes)
+
+    def __repr__(self):
+        return f"{self.__class__.__name__}(data={self._data}, config={self.config})"
+
+    def to_original(self):
+        return self._data.clone()
+
+    # adapated from FP8 implementation of WeightWithDynamicFloat8CastTensor
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs):
+        config = None
+
+        def unwrap(x: cls):
+            nonlocal config
+            if config is None:
+                config = x.config
+            else:
+                assert x.config == config
+            return x._data
+
+        out = func(
+            *pytree.tree_map_only(cls, unwrap, args),
+            **pytree.tree_map_only(cls, unwrap, kwargs),
+        )
+
+        if func is aten.copy_.default:
+            # return original object
+            return args[0]
+        elif func in {
+            aten.t.default,
+            aten.detach.default,
+            aten.empty_like.default,
+            aten.new_zeros.default,
+            aten.slice.Tensor,
+            aten.view.default,
+            aten.as_strided.default,
+            aten._to_copy.default,
+            aten._pin_memory.default,
+            aten.split.Tensor,
+            aten.clone.default,
+        }:
+            # return new wrapped object
+            return pytree.tree_map_only(Tensor, lambda x: cls(x, config), out)
+        else:
+            # return new unwrapped object
+            return out
+
+    def fsdp_pre_all_gather(self, mesh):
```

Comment:
By the way, I was thinking of changing the signature to:
```
def fsdp_pre_all_gather(
    self,
    mesh: DeviceMesh,
    module: nn.Module,  # owning module (e.g. `lin` if this is `lin.weight`)
    mp_policy: MixedPrecisionPolicy,  # policy of the owning FSDP module
):
```
as (1) having access to the `module` can allow using the module as a scratch space for any needed state and (2) having access to the `mp_policy` helps with situations like the one here.

My main concern with doing this is how to deal with the backward compatibility issue. Technically, FSDP2 is still in prototype release, so we are allowed to break BC. However, libraries like ao and torchtune may need to do if/else based on PyTorch version. Maybe it is better to do it sooner than later though.

Reply:
Here we go: https://github.com/pytorch/pytorch/pull/136129
will need more time to test it -- fortunately the subclasses using the FSDP extension are all in ao now it seems 😆